### ISSUE TITLE
Feature request: asofjoin

### ISSUE THREAD

**Issue by natemcintosh** [JuliaData/DataFrames.jl#2738]
This is a request for an `asof` join type, as seen in [pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge_asof.html), [kdb+](https://code.kx.com/q/ref/asof/), [QuestDB](https://questdb.io/docs/reference/sql/join/#asof-join), and probably a few others. 

This is a rather niche join, but extremely useful for time series work. I think the pandas' description sums it up pretty well:

> This is similar to a left-join except that we match on nearest key rather than equal keys. Both DataFrames must be sorted by the key.
> For each row in the left DataFrame:
> - A “backward” search selects the last row in the right DataFrame whose ‘on’ key is less than or equal to the left’s key.
> - A “forward” search selects the first row in the right DataFrame whose ‘on’ key is greater than or equal to the left’s key.
> - A “nearest” search selects the row in the right DataFrame whose ‘on’ key is closest in absolute distance to the left’s key.
>
> The default is “backward” and is compatible in versions below 0.20.0. The direction parameter was added in version 0.20.0 and introduces “forward” and “nearest”.
> Optionally match on equivalent keys with ‘by’ before searching with ‘on’.

**Comment by bkamins** [comment]
OK - request noted. Let us wait for others to comment. Currently most of development is focused on other areas (performance), but we can get back to it in the future.

**Comment by skanskan** [comment]
Is it related with this ?
https://discourse.julialang.org/t/merge-dataframes-where-one-value-is-between-two-others/34738

**Comment by natemcintosh** [comment]
It is very similar, but I believe there is one key difference: the "right" dataframe only has one timestamp, not a start timestamp and an end timestamp. 

A super simple example (without a groupby option) is 

```julia
left = DataFrame(time=[1,5,10], left_val=["a","b","c"])
right = DataFrame(time=[1,2,3,6,7], right_val=[1,2,3,6,7])

# for each time in `left`, find the time (if any) in `right` that 
# is equal to or before it
right_inds = [searchsortedlast(right.time, t) for t in left.time]

# concatenate the dataframes
df = hcat(left, right[right_inds, :], makeunique=true)
```
The pandas version of this function also does forward and nearest selection, but backward selection is probably used most often

**Comment by junder873** [comment]
I agree this function would be very useful. It is definitely a niche to financial data, though I am sure some other fields use it. After asking that question and writing an initial function, I modified it to include a minimize component (which works for asof if the ranges are correct): https://gist.github.com/junder873/ae1ec240f2fd6d4cbdb2ec19bc2b5f9c

I am sure my code is sloppy, so I would love a more formal solution to these types of problems.

**Comment by robsmith11** [comment]
For reference, here are the kdb docs for asof and window joins:
https://code.kx.com/q/ref/aj/
https://code.kx.com/q/ref/wj/

kdb is very well designed and optimized for filtering, joining, and basic data operations on financial market data.  It's definitely worth a look for inspiration for any dataframes library.  I still strongly prefer its syntax vs pandas or Julia's dataframes.

**Comment by natemcintosh** [comment]
> I agree this function would be very useful. It is definitely a niche to financial data, though I am sure some other fields use it. After asking that question and writing an initial function, I modified it to include a minimize component (which works for asof if the ranges are correct): https://gist.github.com/junder873/ae1ec240f2fd6d4cbdb2ec19bc2b5f9c
> 
> I am sure my code is sloppy, so I would love a more formal solution to these types of problems.

I understand how `range_join` performs the join by putting between two other times, but how does it work for an asof join, where you only want a row before or equal to a certain time?

**Comment by junder873** [comment]
It would be using two different components of the function I posted, the conditions component and the minimize component:
```julia
df1 = DataFrame(
    id=1:2,
    date_left=[Date(2020, 1, 5), Date(2020, 5, 3)]
)
df2 = DataFrame(
    id=[1, 1, 1, 2, 2],
    date_right=[Date(2020, 1, 1), Date(2020, 1, 4), Date(2020, 1, 6), Date(2019, 1, 1), Date(2020, 5, 1)],
    val=1:5
)



range_join(
    df1,
    df2,
    [:id],
    Tuple{Function, Symbol, Symbol}[
        (>, :date_left, :date_right)
    ],
    minimize=[:date_left => :date_right]
)
```

**Comment by bkamins** [comment]
So I would propose the following (we would need to think about the best name though; technically we could add this extension to `leftjoin` as it is not breaking, but I was not sure if it is not better to add a new name as the functionality is more complex, although `leftjoin` is a special case of it; potentially we could add this API to all joins but I am not sure if it is worth it - let us discuss):
```
rulejoin(df1, df2; on, condition=(x...) -> all(t -> isequal(t[1], t[2]), x), makeunique=false, source=nothing, validate=(false, false), renamecols=(identity => identity), window=nothing, keep=:all)
```
This would behave like `leftjoin` except that new `condition`, `window` and `keep` kwargs are added.

* `condition`: a function taking as many positional arguments as are `on` columns, each positional argument is a tuple of values from left and right table; by default it checks if values from left and right table are equal (so this is the default `leftjoin`); we could add predefined functions for some standard conditions (like maximum absolute distance, greater than etc.)
* `keep`: which matching rows from the right table to keep. By default `:all` to keep all rows. Other options are `:first` - keep first matching row, `:last` - keep last matching row, or a function, in this case it is taking as many positional arguments as are `on` columns, each positional argument is a tuple of values from left and right table and it should return values that are comparable and have total order; the returned value is the first row that is a minimal element of the set of matching rows;
* `window`:
    * if `nothing` then the operation is O(n*m) - just all combinations of left and right table are checked; the result is produced in the order of rows from `df1`
    * if `:ascending` then `df1` and `df2` are sorted before operation and it is guaranteed by the user that `condition` is a "window function" (i.e. as you move along the condition after data frames are sorted the matching rows are a continuous block and both begin and end of the block are increasing); in this case, after sorting the cost of matching is O(n+m). Sorting is done only if needed in particular (so 
    * if `:descending` then the behavior is the same as for `:ascending`, but with `rev=true` when sorting

Adding this feature will cover all kinds of joins on arbitrary condition, with all kinds of "as of" and similar joins being a special case and fast.

What I am not sure about if it is not overly general (and thus not needed in practice). Maybe it is enough to ensure such matching only on a single column and just allow other `on` columns to be standard equijoin conditions?

CC @nalimilan

**Comment by bkamins** [comment]
@jariji - this is another important PR where API design is a key challenge.

**Comment by jariji** [comment]
I'll just mention a related thing for now. Sometimes I want fuzzy joins like `on=:birthdate, nearest=:name => levenshtein` or `on=:birthdate, filter=:name => <(4)∘levenshtein`

**Comment by bkamins** [comment]
I was thinking of "nearest" join, but could not decide what to do with it (current proposal does not cover nearest join).
If this is indeed needed, we could change `condition` from requiring to return `Bool` (is condition met) to `distance` and then add option `threshold` (or something like this), where only pairs that have `distance` less than `threshold` are considered as matching, and one of the options for `threshold` would be e.g. `min` (to select only nearest match).

**Comment by robsmith11** [comment]
I've used this package for various time series joins:
https://gitlab.com/aplavin/FlexiJoins.jl

The syntax takes a little getting used to, but it's pretty powerful and performance is decent, whether joining DataFrames or other types.

**Comment by kdpsingh** [comment]
Thanks! That's super helpful. Quick question: what is the purpose of `identity` in the `multi` argument?

I think in that case my only critique of the FlexiJoins.jl implementation is that when matching on name (or equality by pred), there's no need to create `a` and `a_1`. For inequality joins, it makes sense to have both (since they won't always match), but otherwise it creates an extra column that's not needed, especially when they have the same name to begin with.

Otherwise, I would be happy if this were incorporated as core functionality within DataFrames.jl. It's certainly less verbose than what I proposed.

**Comment by robsmith11** [comment]
> what is the purpose of identity in the multi argument?

My understanding is that it's just a placeholder, so that it's clear to which table the `closest` applies.  For other data types, it's possible to pass `multi` a named tuple and then `identity` isn't needed, but for some reason that I forget, that won't work with DataFrames.

**Comment by kdpsingh** [comment]
Thanks @aplavin! Appreciate the answers. None of what I said was intended as criticism - it's more stuff that DataFrames.jl maintainers should consider as they think about the best way to support flexible joins using DataFrames. I totally recognize that your package works with a much broader set of data structures.
